be/src/storage/column_predicate_rewriter.cpp

```diff
@@ -48,46 +48,121 @@ namespace starrocks {
 
 constexpr static const LogicalType kDictCodeType = TYPE_INT;
 
-Status ColumnPredicateRewriter::rewrite_predicate(ObjectPool* pool) {
-    // because schema has reordered
-    // so we only need to check the first `predicate_column_size` fields
-    for (size_t i = 0; i < _column_size; i++) {
-        const FieldPtr& field = _schema.field(i);
-        const auto cid = field->id();
+struct RewritePredicateTreeVisitor {
+    using RewriteStatus = ColumnPredicateRewriter::RewriteStatus;
+
+    template <CompoundNodeType ParentType>
+    StatusOr<RewriteStatus> operator()(PredicateColumnNode& node, PredicateCompoundNode<ParentType>& parent) const {
+        const auto* col_pred = node.col_pred();
+        const auto cid = col_pred->column_id();
 
-        auto iter = _pred_map.find(cid);
-        if (iter == _pred_map.end()) {
-            continue;
+        if (!_rewriter._need_rewrite[cid]) {
+            return RewriteStatus::UNCHANGED;
         }
-        auto& preds = iter->second;
 
-        std::vector<const ColumnPredicate*> remove_list;
-        for (auto& pred : preds) {
-            ColumnPredicate* new_pred = nullptr;
-            ASSIGN_OR_RETURN(auto rewrite_status, _rewrite_predicate(pool, field, pred, &new_pred));
+        const auto& field = _cid_to_field.find(cid)->second;
+        DCHECK(_rewriter._column_iterators[cid]->all_page_dict_encoded());
+
+        ColumnPredicate* rewrited_pred;
+        ASSIGN_OR_RETURN(auto rewrite_status, _rewriter._rewrite_predicate(_pool, field, col_pred, &rewrited_pred));
+
+        if (rewrite_status == RewriteStatus::CHANGED) {
+            _pool->add(rewrited_pred);
+            parent.add_child(PredicateColumnNode{rewrited_pred});
+        }
+
+        return rewrite_status;
+    }
+
+    template <CompoundNodeType Type, CompoundNodeType ParentType>
+    StatusOr<RewriteStatus> operator()(PredicateCompoundNode<Type>& node,
+                                       PredicateCompoundNode<ParentType>& parent) const {
+        std::vector<PredicateNodePtr> unchanged_children;
+        unchanged_children.reserve(node.num_children());
+        auto new_node = PredicateCompoundNode<Type>{};
+
+        bool changed = false;
+        for (auto child : node.children()) {
+            ASSIGN_OR_RETURN(auto rewrite_status, child.visit(*this, new_node));
+
+            changed |= rewrite_status != RewriteStatus::UNCHANGED;
 
             switch (rewrite_status) {
             case RewriteStatus::ALWAYS_TRUE:
-                remove_list.emplace_back(pred);
-                break;
+                if constexpr (Type == CompoundNodeType::AND) {
+                    break; // Do nothing.
+                } else {
```

Comment:
here else-branch means parent node is a OR predicates?

Reply:
No, it means node itself is a OR predicate.

For example, as for `c1='a' OR c2='b'`, `node` is OR predicate, children is `c1='a'` and `c2='b'`.
If `c1='a'` is a always true predicate, then it can just remove `c1='a'1, and be simplified from `c1='a' OR c2='b'` to `c2='b'`.

Reply:
`ParentType` is just used to pass `PredicateCompoundNode<ParentType>& parent`, we don't use `ParentType` in this function.